Refactor typing

Goldziher opened the pull request:
This PR refactors the typing helpers we use - merging several different wrappers into `ParsedType`. 

This is a rather complex refactor because we had several different competing idioms in place, but its necessary exactly because of this reason. 

Remaining- Fix about 200 tests 😉

JacobCoffee commented:
> Fix about 200 tests 😉

._.

Goldziher commented on docs/usage/dependency-injection.rst:
```diff
@@ -364,7 +364,7 @@ Depending on your application design, it is possible to have a dependency declar
 default should be used by the function.
 
-.. literalinclude:: /examples/dependency_injection/dependency_default_value_no_dependency_fn.py
+.. literalinclude:: /examples/dependency_injection/dependency_default_value_with_dependency_fn.py
```
Wrong name, should rename file bk

JacobCoffee commented on litestar/params.py:
```diff
@@ -106,6 +106,10 @@ class KwargDefinition:
     upper_case: bool | None = field(default=None)
     """Constrict a string value to be upper case."""
+    format: str | None = field(default=None)
+    """Constrict a string value to be upper case."""
```
```suggestion
    """Specify the format to which a string value should be converted."""
```

JacobCoffee reviewed (changes requested):
Sorry if dumb question - What does the change give us here?

Goldziher commented:
> Sorry if dumb question - What does the change give us here?

Read the PR desc

Goldziher commented:
> Sorry if dumb question - What does the change give us here?

Maybe I missed what you are asking, you mean the format stuff added on the ParameterKwarg? Its there to allow it being copied over from metadata. Specifically from Piccolo metadata.

peterschutt commented:
> Merge `ParsedParameter` info `ParsedType`

This doesn't make a lot of sense to me - the two have different properties. Parameters have a name, type and default value, and a type is just a type.

Goldziher commented:
> > Merge `ParsedParameter` info `ParsedType`
> 
> This doesn't make a lot of sense to me - the two have different properties. Parameters have a name, type and default value, and a type is just a type.

It's true it's not optimal BUT consider the following:

1. Types can contain a lot more data because they can contain metadata via annotation. In fact this is becoming the de facto standard (especially for us), which actually includes the default value and a whole bunch of other stuff. 

2. We have a lot of redundancy and deep nesting, with duplication of logic. Deduplicating the logic means adding a name field, which is potentially empty. 

Saying that, we can split 🪓 `ParsedType` and move some of the logic into `FieldDefinition` - or reintroduce `ParsedParameter` as a subclass of `ParsedType` and superclass of `FieldDefinition`. 

This might we'll be cleaner. 

We should also consider whether this stuff should be private namespaces, since typing is a public module.

gsakkis commented on litestar/typing.py:
```diff
@@ -165,3 +385,135 @@ def has_inner_subclass_of(self, cl: type[Any] | tuple[type[Any], ...]) -> bool:
+        inner_types: tuple[FieldDefinition, ...] | None = None,
+        kwarg_definition: KwargDefinition | DependencyKwarg | None = None,
+        extra: dict[str, Any] | None = None,
+    ) -> FieldDefinition:
```
This looks like a restricted wrapper of `from_annotation`. Is it really needed? How does one choose which to use?

gsakkis commented on litestar/typing.py:
```diff
@@ -42,56 +175,142 @@ class ParsedType:
+        return hash((self.name, self.raw, self.annotation, self.origin, self.inner_types))
+
+    @classmethod
+    def _extract_metada(
```
`_extract_metadata`?

Goldziher replied:
yup, will fix

gsakkis reviewed (commented):
I skimmed over most of it but lgtm overall, a few minor comments/questions.

Goldziher commented on litestar/typing.py:
```diff
@@ -165,3 +385,135 @@ def has_inner_subclass_of(self, cl: type[Any] | tuple[type[Any], ...]) -> bool:
+        return FieldDefinition.from_kwarg(
+            annotation=annotation,
+            name=parameter.name,
+            default=Empty if parameter.default is Signature.empty else parameter.default,
```
they are, but i dont see the benefit of doing this. what would it be?

gsakkis replied:
"There should be one-- and preferably only one --obvious way to do it". Now there are two (or three if you add the Ellipsis)

Goldziher commented on litestar/typing.py:
```diff
@@ -165,3 +385,135 @@ def has_inner_subclass_of(self, cl: type[Any] | tuple[type[Any], ...]) -> bool:
+                    "extra": extra,
+                    "default": default,
+                }.items()
+                if v
```
fixed

Goldziher commented on litestar/typing.py:
```diff
@@ -42,56 +175,142 @@ class ParsedType:
+            return _create_metadata_from_type(metadata=[annotation], model=model, annotation=annotation, extra=extra)
+
+        if any(isinstance(arg, KwargDefinition) for arg in get_args(annotation)):
+            return [arg for arg in get_args(annotation) if isinstance(arg, KwargDefinition)][0], extra or {}
```
addressed

Goldziher commented on litestar/typing.py:
```diff
@@ -42,56 +175,142 @@ class ParsedType:
+            return _create_metadata_from_type(metadata=[annotation], model=model, annotation=annotation, extra=extra)
+
+        if isinstance(annotation, AbstractDTOFactory):
+            return _create_metadata_from_type(metadata=[annotation], model=model, annotation=annotation, extra=extra)
```
fixed

Goldziher commented on docs/release-notes/changelog.rst:
fixed

provinzkraut commented:
> Deduplicating the logic means adding a name field, which is potentially empty. The cost of this is very minimal.

So, I think the general idea is a good one. However, I do also agree with @peterschutt that the original distinction between the representation of a parsed type and that of a parameter is a valuable one.

Maybe there is a different way to approach this issue?

I already looked into doing something similar while working on #1145, and I think with that, we might simply be able to *remove* a lot of the abstractions, since they're not needed anymore. That could accomplish a similar goal I think?

Goldziher commented:
> Member

As i wrote above - i removed the `ParsedType` and renamed this to `FieldDefinition`. Our use case is not for types as much as for actual fields, and we use this over the entire place (just see the places this PR touches). Furthermore, there is a whole lot of duplication. The only difference is between a value having a name, and not having a name. I do not think we need to have the complexity for an optional string value.

cofin reviewed (approved):
I feel like I've accomplished something by just reading through this one.  This is massive, but it looks good to me.
